### ISSUE TITLE
Stop doing DVE-2018-0001 workaround by default

### ISSUE THREAD

**Issue by pemensik** [systemd/systemd#26967]
### systemd version the issue has been seen with

253.1

### Used distribution

Fedora Rawhide

### Linux kernel version used

6.1.10-100.fc36.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-resolved

### Expected behaviour you didn't see

I did 3 queries on non-existent names, but 6 is forwarded when DNSSEC=no (default). I expected only one for each query.

```
dig @127.0.0.53 nx{1,2,3}.example.net
```

### Unexpected behaviour you saw

```
$ tcpdump -nvr resolved-DVE-2018-0001.pcapng 
reading from file resolved-DVE-2018-0001.pcapng, link-type EN10MB (Ethernet), snapshot length 262144

14:51:19.897971 IP (tos 0x0, ttl 64, id 60553, offset 0, flags [none], proto UDP (17), length 72)
    192.168.122.184.33302 > 192.168.122.1.domain: 6611+ [1au] A? nx1.example.net. (44)
14:51:19.898371 IP (tos 0x0, ttl 64, id 53775, offset 0, flags [DF], proto UDP (17), length 134)
    192.168.122.1.domain > 192.168.122.184.33302: 6611 NXDomain 0/1/1 (106)
14:51:19.898497 IP (tos 0x0, ttl 64, id 60554, offset 0, flags [none], proto UDP (17), length 61)
    192.168.122.184.33302 > 192.168.122.1.domain: 6611+ A? nx1.example.net. (33)
14:51:19.898568 IP (tos 0x0, ttl 64, id 53776, offset 0, flags [DF], proto UDP (17), length 61)
    192.168.122.1.domain > 192.168.122.184.33302: 6611 NXDomain 0/0/0 (33)
14:51:19.899906 IP (tos 0x0, ttl 64, id 38244, offset 0, flags [none], proto UDP (17), length 72)
    192.168.122.184.35392 > 192.168.122.1.domain: 4701+ [1au] A? nx2.example.net. (44)
14:51:19.900193 IP (tos 0x0, ttl 64, id 53777, offset 0, flags [DF], proto UDP (17), length 134)
    192.168.122.1.domain > 192.168.122.184.35392: 4701 NXDomain 0/1/1 (106)
14:51:19.900351 IP (tos 0x0, ttl 64, id 38245, offset 0, flags [none], proto UDP (17), length 61)
    192.168.122.184.35392 > 192.168.122.1.domain: 4701+ A? nx2.example.net. (33)
14:51:19.900409 IP (tos 0x0, ttl 64, id 53778, offset 0, flags [DF], proto UDP (17), length 61)
    192.168.122.1.domain > 192.168.122.184.35392: 4701 NXDomain 0/0/0 (33)
14:51:19.901445 IP (tos 0x0, ttl 64, id 59486, offset 0, flags [none], proto UDP (17), length 72)
    192.168.122.184.49224 > 192.168.122.1.domain: 1832+ [1au] A? nx3.example.net. (44)
14:51:19.911710 IP (tos 0x0, ttl 64, id 53779, offset 0, flags [DF], proto UDP (17), length 128)
    192.168.122.1.domain > 192.168.122.184.49224: 1832 NXDomain 0/1/1 (100)
14:51:19.911799 IP (tos 0x0, ttl 64, id 59487, offset 0, flags [none], proto UDP (17), length 61)
    192.168.122.184.49224 > 192.168.122.1.domain: 1832+ A? nx3.example.net. (33)
14:51:19.911854 IP (tos 0x0, ttl 64, id 53780, offset 0, flags [DF], proto UDP (17), length 61)
    192.168.122.1.domain > 192.168.122.184.49224: 1832 NXDomain 0/0/0 (33)
```

It retries query after EVERY query with edns again without it. For all good servers this is wrong and wasting connectivity and CPU cycles. Just to workaround broken implementation violating RFC 1035.

### Steps to reproduce the problem

- ensure DNSSEC=no in resolved.conf
- start recording of putgoing dns queries.
- run command ``dig @127.0.0.53 nx{1,2,3}.example.net``
- check how many queries were forwarded

### Additional program output to the terminal or log subsystem illustrating the issue

It even repeats responses again, even if it could have fresh and cached negative answer. Most good resolvers provide SOA in NXDOMAIN reply, which specifies for how long the negative answer should be cached. That is not supported.

But anyway, [DVE-2018-0001](https://github.com/dns-violations/dns-violations/blob/master/2018/DVE-2018-0001.md) workaround should *not* be active for all users. Certainly not in current way, where it duplicates every negative response for every nxdomain answer. Apparently it increases queries number for all well behaved servers. Just because some implementors made a mistake in their product, where it should be fixed *only*. If it has to detect such misfeature, it has to cache it for some time, ideally until name servers or interfaces do change. I think this workaround should be active only after manual configuration, even if it would be cached. By default it should not attempt it at all. Currently it is not possible to disable it.

**Comment by pemensik** [comment]
This is related to PR #18638 introducing this workaround.

**Comment by pemensik** [comment]
I think it would be better if Network Manager's option ``ipv4.dns-options`` could be used to omit edns0 just for this specific network. Not sure network manager can pass those options to resolved. It should be just rare workaround needed on some networks, not for all of them.

**Comment by xnox** [comment]
My preference is to revert all of https://github.com/systemd/systemd/pull/18638/files

**Comment by xnox** [comment]
> I think it would be better if Network Manager's option `ipv4.dns-options` could be used to omit edns0 just for this specific network. Not sure network manager can pass those options to resolved. It should be just rare workaround needed on some networks, not for all of them.

In resolved, internally settings are on per-server-per-connection basis - rather than on per-connection basis. And feature auto-detection happens automatically. Meaning if there are no responses with EDNS0 and DNSSEC is not enforced, resolved will cycle down to using more simple queries for a given DNS server. So there shouldn't be an explicit need to set such options from network-manager to systemd-resolved.

Passing other settings from network-manager to systemd-resolved might make sense - i.e. enforced/opportunistic DNSSEC, DNS over TLS. But these are global, rather than per-server-per-connection settings.

**Comment by xnox** [comment]
> I do not ask them per-server. I ask them per interface and all their servers, just like _dns-over-tls_ setting now allows. I think that would be enough for majority of users.

all of the above is unrelated to removing this old workaround for Aruba networks. Please open separate issue if you want more things to set/integrate between resolved and networkmanager. Still failing to understand why additional requests are being piled into this issue. The two are not related, and this workaround was never added to support explicit featurelevel configuration and removing this workaround shouldn't be blocked on additional new featurelevel configurations either.

**Comment by xnox** [comment]
I also don't see any evidence if additional per-interface settings are desired, required, or useful. Just because we can make everything and anything configurable, doesn't mean we should.
